<a href="https://colab.research.google.com/github/ahdrifai1234/data-science-2026/blob/main/Pertemuan12_Ahmad_Rifai_240401010260.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

IDENTITAS

NAMA  : Ahmad Rifai

NIM   : 240401010260

KELAS : IF403

In [4]:
# Langkah 1: Generate & Eksplorasi Dataset Transaksi
import pandas as pd, numpy as np
import matplotlib.pyplot as plt
np.random.seed(42)
produk = ['Roti', 'Selai', 'Susu', 'Sereal', 'Telur',
'Keju', 'Kopi', 'Gula', 'Teh', 'Mentega']
# Buat 50 transaksi, tiap transaksi berisi 2-5 produk
transaksi = []
for _ in range(50):
  n_item = np.random.randint(2, 6)
  transaksi.append(list(np.random.choice(produk, n_item, replace=False)))
# Suntikkan pola: Roti sering bersama Selai
for i in range(0, 20):
  if 'Roti' in transaksi[i] and 'Selai' not in transaksi[i]:
    transaksi[i].append('Selai')

print('Contoh transaksi:', transaksi[:3])
print('Jumlah transaksi:', len(transaksi))

Contoh transaksi: [[np.str_('Keju'), np.str_('Roti'), np.str_('Mentega'), np.str_('Kopi'), 'Selai'], [np.str_('Roti'), np.str_('Kopi'), np.str_('Teh'), np.str_('Selai'), np.str_('Mentega')], [np.str_('Kopi'), np.str_('Susu'), np.str_('Teh')]]
Jumlah transaksi: 50


In [5]:
# Langkah 2: One-Hot Encoding Transaksi
from mlxtend.preprocessing import TransactionEncoder
te = TransactionEncoder()
te_ary = te.fit(transaksi).transform(transaksi)
df = pd.DataFrame(te_ary, columns=te.columns_)
print(df.head())

    Gula   Keju   Kopi  Mentega   Roti  Selai  Sereal   Susu    Teh  Telur
0  False   True   True     True   True   True   False  False  False  False
1  False  False   True     True   True   True   False  False   True  False
2  False  False   True    False  False  False   False   True   True  False
3  False   True  False    False  False   True   False  False   True   True
4   True   True  False     True  False  False   False   True  False  False


In [6]:
# Langkah 3: Cari Frequent Itemset dengan Apriori
from mlxtend.frequent_patterns import apriori
for ms in [0.05, 0.1, 0.2]:
  freq = apriori(df, min_support=ms, use_colnames=True)
  print(f'min_support={ms}: {len(freq)} itemset ditemukan')

# Gunakan min_support yang menghasilkan jumlah itemset wajar (tidak 0, tidak ratusan)
freq_items = apriori(df, min_support=0.1, use_colnames=True)
freq_items = freq_items.sort_values('support', ascending=False)
print(freq_items.head(10))

min_support=0.05: 74 itemset ditemukan
min_support=0.1: 44 itemset ditemukan
min_support=0.2: 13 itemset ditemukan
    support      itemsets
5      0.52       (Selai)
8      0.46         (Teh)
3      0.42     (Mentega)
9      0.36       (Telur)
1      0.34        (Keju)
0      0.32        (Gula)
2      0.32        (Kopi)
4      0.32        (Roti)
7      0.32        (Susu)
36     0.24  (Teh, Selai)


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [7]:
# Langkah 4: Bentuk & Saring Aturan Asosiasi
from mlxtend.frequent_patterns import association_rules
rules = association_rules(freq_items, metric='confidence',
min_threshold=0.5)
rules = rules[rules['lift'] > 1].sort_values('lift', ascending=False)
print(rules[['antecedents', 'consequents',
'support', 'confidence', 'lift']].head(10))

         antecedents consequents  support  confidence      lift
9        (Keju, Teh)     (Telur)     0.12    0.857143  2.380952
13  (Selai, Mentega)      (Kopi)     0.10    0.625000  1.953125
12      (Gula, Roti)     (Selai)     0.10    1.000000  1.923077
7           (Sereal)   (Mentega)     0.14    0.777778  1.851852
10      (Telur, Teh)      (Keju)     0.12    0.600000  1.764706
14     (Selai, Kopi)   (Mentega)     0.10    0.714286  1.700680
8      (Keju, Telur)       (Teh)     0.12    0.750000  1.630435
11     (Gula, Selai)      (Roti)     0.10    0.500000  1.562500
15   (Mentega, Kopi)     (Selai)     0.10    0.714286  1.373626
1             (Roti)     (Selai)     0.22    0.687500  1.322115


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

### Interpretasi Aturan Asosiasi

Dari hasil aturan asosiasi yang diurutkan berdasarkan nilai `lift` tertinggi:

1.  **Aturan Terkuat (Lift Tertinggi):**
    *   `antecedents={'Keju', 'Teh'}`, `consequents={'Telur'}` dengan `lift` **2.38**. Ini menunjukkan bahwa pelanggan yang membeli Keju dan Teh sangat mungkin juga membeli Telur, dan kombinasi ini muncul lebih dari 2 kali lipat dari yang diperkirakan secara acak.

2.  **Aturan Roti -> Selai:**
    *   Terdapat aturan `antecedents={'Gula', 'Roti'}`, `consequents={'Selai'}` dengan `lift` **1.92**. Ini sangat masuk akal secara bisnis, karena Roti dan Selai adalah kombinasi sarapan yang umum, dan adanya Gula mungkin menandakan preferensi untuk rasa manis.
    *   Juga ada aturan `antecedents={'Roti'}`, `consequents={'Selai'}` dengan `lift` **1.32**. Ini mengkonfirmasi pola umum bahwa pembeli Roti cenderung membeli Selai, sesuai dengan ekspektasi bisnis.

Secara umum, aturan-aturan ini memberikan wawasan yang berguna untuk strategi penempatan produk, promosi bundling, atau rekomendasi kepada pelanggan.

In [10]:
# Langkah 5: Rekomender Sederhana dengan Content-Based Filtering
from sklearn.metrics.pairwise import cosine_similarity
katalog = pd.DataFrame({
'produk': produk,
'kategori': ['Bakery','Bakery','Dairy','Bakery','Dairy',
'Dairy','Minuman','Bumbu','Minuman','Dairy']
})
fitur = pd.get_dummies(katalog['kategori'])
sim_matrix = cosine_similarity(fitur)

def rekomendasi_serupa(nama_produk, top_n=3):
  idx = katalog.index[katalog['produk'] == nama_produk][0]
  skor = list(enumerate(sim_matrix[idx]))
  skor = sorted(skor, key=lambda x: x[1], reverse=True)
  skor = [s for s in skor if s[0] != idx][:top_n]
  return katalog.iloc[[i for i, _ in skor]]['produk'].tolist()
print('Mirip dengan Roti:', rekomendasi_serupa('Roti'))

Mirip dengan Roti: ['Selai', 'Sereal', 'Susu']


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [11]:
# Langkah 6: Bandingkan Kedua Pendekatan
produk_target = 'Roti'
# Dari association rules: cari consequents dari aturan yang antecedent-nya mengandung
produk_target
rules_terkait = rules[rules['antecedents'].apply(
lambda x: produk_target in x)]
print('Rekomendasi dari Association Rules:')
print(rules_terkait[['consequents', 'lift']].head())
print('Rekomendasi dari Content-Based:', rekomendasi_serupa(produk_target))

Rekomendasi dari Association Rules:
   consequents      lift
12     (Selai)  1.923077
1      (Selai)  1.322115
Rekomendasi dari Content-Based: ['Selai', 'Sereal', 'Susu']


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

### Diskusi Perbandingan Kedua Pendekatan

**Konsistensi Rekomendasi:**
*   **Rekomendasi dari Association Rules untuk 'Roti':** Memberikan 'Selai' sebagai *consequent* dengan lift tinggi (1.92 dan 1.32), yang menunjukkan hubungan kuat berdasarkan perilaku pembelian bersama. Ini sangat masuk akal secara bisnis.
*   **Rekomendasi dari Content-Based untuk 'Roti':** Memberikan 'Selai', 'Sereal', dan 'Susu'. 'Selai' juga direkomendasikan karena memiliki kategori yang sama ('Bakery') dengan 'Roti'. 'Sereal' dan 'Susu' juga mungkin direkomendasikan karena kesamaan kategori (misalnya, 'Sereal' juga bisa masuk kategori 'Bakery' atau sarapan, dan 'Susu' sering dikonsumsi bersama).

Terlihat ada **konsistensi** dalam merekomendasikan 'Selai' untuk 'Roti' di kedua pendekatan, yang memperkuat keyakinan bahwa ini adalah rekomendasi yang relevan. Pendekatan Content-Based juga memberikan rekomendasi tambahan yang secara kontekstual mirip ('Sereal', 'Susu').

**Kapan Menggunakan Masing-masing atau Menggabungkannya:**

*   **Association Rules:**
    *   **Kapan:** Sangat efektif ketika ingin menemukan hubungan produk berdasarkan *perilaku pembelian aktual* dari banyak pelanggan (misalnya, "apa yang sering dibeli bersama?"). Cocok untuk *cross-selling* dan penempatan produk.
    *   **Kelebihan:** Menemukan pola tersembunyi yang mungkin tidak terdeteksi hanya dari atribut produk. Tidak memerlukan informasi konten produk.
    *   **Kekurangan:** Membutuhkan data transaksi yang cukup banyak. Sulit diterapkan untuk produk baru (masalah *cold start*).

*   **Content-Based Filtering:**
    *   **Kapan:** Cocok ketika ingin merekomendasikan produk yang *serupa secara atribut* dengan produk yang disukai pelanggan (misalnya, "jika Anda menyukai X, Anda mungkin menyukai Y yang serupa"). Berguna untuk mengeksplorasi variasi produk dalam kategori yang sama.
    *   **Kelebihan:** Bekerja dengan baik untuk produk baru jika atributnya diketahui. Mudah dijelaskan mengapa suatu produk direkomendasikan (karena kesamaan atribut).
    *   **Kekurangan:** Rekomendasi bisa jadi kurang beragam (hanya merekomendasikan apa yang sudah mirip). Membutuhkan informasi atribut produk yang detail.

*   **Pendekatan Hybrid (Menggabungkan Keduanya):**
    *   **Kapan:** Seringkali menjadi solusi terbaik. Menggabungkan kekuatan kedua pendekatan dapat menghasilkan rekomendasi yang lebih akurat, relevan, dan beragam.
    *   **Cara Menggabungkan:**
        *   **Weighted Hybrid:** Memberi bobot pada skor dari setiap sistem rekomendasi.
        *   **Feature Combination:** Menggabungkan fitur dari kedua metode sebelum melatih model rekomendasi.
        *   **Switching Hybrid:** Menggunakan satu metode ketika kondisi tertentu terpenuhi, dan metode lain dalam kondisi lain (misalnya, Content-Based untuk cold start, lalu beralih ke Collaborative Filtering/Association Rules setelah ada data interaksi).

Dalam kasus ini, kombinasi keduanya akan sangat kuat. Association Rules dapat mengidentifikasi item yang sering dibeli bersama, sementara Content-Based dapat memperluas rekomendasi ke item lain yang serupa secara kategoris, bahkan jika belum ada pola pembelian bersama yang kuat. Misalnya, jika 'Roti' dan 'Selai' sering dibeli bersama (Association Rules), Content-Based bisa merekomendasikan jenis roti atau selai lain yang belum pernah dicoba pelanggan.

## Kesimpulan Akhir

Analisis ini berhasil mengidentifikasi pola pembelian pelanggan melalui **Association Rules**, menunjukkan item yang sering dibeli bersama (misalnya, Roti dan Selai). Selain itu, **Content-Based Filtering** memberikan rekomendasi produk serupa berdasarkan atribut kategori. Kedua pendekatan ini, meskipun dengan cara berbeda, mengkonfirmasi rekomendasi yang relevan dan dapat digabungkan (hybrid) untuk sistem rekomendasi yang lebih kuat, memberikan wawasan untuk strategi penjualan dan penempatan produk yang efektif.